sycamore/evaluation/pipeline.py

```diff
@@ -1,116 +1,143 @@
+                                    "query_text": doc["question"],
+                                    "model_id": self._os_config["embedding_model_id"],
+                                    "k": self._os_config["neural_search_k"],
+                                    "filter": [{"exists": {"field": "parent_id"}}],
                                 }
                             }
-                        ],
-                        "filter": [{"exists": {"field": "parent_id"}}],
-                    }
+                        },
+                    ]
+                }
+            },
+            "size": self._os_config["size"],
+        }
+
+        if "llm" in self._os_config:
+            query["params"] = {"search_pipeline": self._os_config["search_pipeline"]}
+            query["query"]["ext"] = {
+                "generative_qa_parameters": {
+                    "llm_question": doc["question"],
+                    "context_size": self._os_config["context_window"],
+                    "llm_model": self._os_config["llm"],
                 },
-                "size": self._os_config["size"],
+                "rerank": {"query_context": {"query_text": doc["question"]}},
             }
+        if "filters" in doc:
+            query["query"] = self._add_filter(query["query"], doc["filters"])
         return query
 
-    def _evaluate_queries(self, query_result: Document) -> Document:
+    def _process_queries(self, query_result: Document) -> Document:
+        logger.debug(("Query result: " + str(query_result.keys())))
         assert query_result.type == "OpenSearchQueryResult"
         result = EvaluationDataPoint(query_result)
+        result.generated_source_documents = [Element(hit) for hit in query_result["hits"]]
         metrics = {}
         for metric in self._metrics:
             metrics[metric.metric_name()] = metric.evaluate(result)
         result["metrics"] = metrics
         return result
 
-    def execute(self, input_dataset: DocSet) -> Tuple[DocSet, DocSet]:
+    def _aggregate_metrics(self, query_level_metrics: DocSet) -> Document:
+        """
+        Currently this method averages all metric types for each document. For metrics that need to be handled
+        differently we will need to extend this.
+        """
+        # todo: move these to DocSet level operations
+        summary = EvaluationSummary()
+        summary.metrics = {}
+        metric_data: dict[str, Any] = {}
+
+        for metric in self._metrics:
+            metric_data[metric.metric_name()] = {}
+            summary.metrics[metric.metric_name()] = {}
+
+        for doc in query_level_metrics.take_all():
+            for metric_key, metric_value in metric_data.items():
+                for sub_metric_key, sub_metric_value in doc["metrics"].get(metric_key, {}).items():
+                    metric_data[metric_key][sub_metric_key] = metric_data[metric_key].get(sub_metric_key, []) + [
+                        sub_metric_value
+                    ]
```

Comment:
there's gotta be a fancy ray mapreduce-y/spark-y way to do this

Reply:
Yeah I have some references from Ben how to do a group/aggregate op but it was a little complex/overkill for this case. Will evaluate later